src/System.Management.Automation/engine/lang/scriptblock.cs

```diff
@@ -647,16 +647,17 @@ internal ReadOnlyCollection<PSTypeName> OutputType
         /// <remarks>
         /// This does normal array reduction in the case of a one-element array.
         /// </remarks>
-        internal static object GetRawResult(List<object> result)
+        internal static object GetRawResult(List<object> result, bool wrapToPSObject)
         {
             switch (result.Count)
             {
                 case 0:
                     return AutomationNull.Value;
                 case 1:
-                    return LanguagePrimitives.AsPSObjectOrNull(result[0]);
+                    return wrapToPSObject ? LanguagePrimitives.AsPSObjectOrNull(result[0]) : result[0];
                 default:
-                    return LanguagePrimitives.AsPSObjectOrNull(result.ToArray());
+                    object resultArray = result.ToArray();
+                    return wrapToPSObject ? LanguagePrimitives.AsPSObjectOrNull(resultArray) : resultArray;
```

Comment:
Is the converting to immutable array mandatory?

Reply:
@iSazonov Sorry for the super long delay to respond to your comment.
The `result.ToArray()` is existing code, and it would be a breaking change  to end users to change that because users may expect an object array as it's always be.

Reply:
If the PR breaks an C# application in any case and developers have to fix their code we could weight removing `result.ToArray()` too. If we think about enumeration the replacing List<object> with Array does seem not break this.

Reply:
@daxian-dbw Could you please share your thoughts about my last comment?

Reply:
Note that this method is also used by `DoInvokeReturnAsIs`. The current change keeps the existing behavior of that API, but removing `result.ToArray()` definitely breaks it.
Also, there is no obvious benefit by removing `result.ToArray()` except for a bit less allocation that may not be noticeable, so I don't think a breaking change can be justified.